olp-cpp-sdk-dataservice-read/src/repositories/PrefetchTilesRepository.cpp

```diff
@@ -177,56 +225,37 @@ SubQuadsResponse PrefetchTilesRepository::GetSubQuads(
     return query_api.GetError();
   }
 
-  using QuadTreeIndexResponse = QueryApi::QuadTreeIndexResponse;
-  using QuadTreeIndexPromise = std::promise<QuadTreeIndexResponse>;
-  auto promise = std::make_shared<QuadTreeIndexPromise>();
   auto tile_key = tile.ToHereTile();
 
   OLP_SDK_LOG_INFO_F(kLogTag,
                      "GetSubQuads execute(%s, %" PRId64 ", %" PRId32 ")",
                      tile_key.c_str(), version, depth);
 
   auto quad_tree = QueryApi::QuadTreeIndex(
-      query_api.GetResult(), layer_id, version, tile_key, depth, boost::none,
+      query_api.GetResult(), layer_id, tile_key, version, depth, boost::none,
       request.GetBillingTag(), context);
 
-  if (!quad_tree.IsSuccessful()) {
+  if (quad_tree.status != olp::http::HttpStatusCode::OK) {
     OLP_SDK_LOG_WARNING_F(kLogTag,
                           "GetSubQuads failed(%s, %" PRId64 ", %" PRId32 ")",
                           tile_key.c_str(), version, depth);
-    return quad_tree.GetError();
+    return {{quad_tree.status, quad_tree.response.str()}};
   }
 
-  SubQuadsResult result;
-  model::Partitions partitions;
-
-  const auto& subquads = quad_tree.GetResult().GetSubQuads();
-  partitions.GetMutablePartitions().reserve(subquads.size());
-
-  OLP_SDK_LOG_DEBUG_F(kLogTag,
-                      "GetSubQuad finished, key=%s, size=%zu,  version=%" PRId64
-                      ", depth=%" PRId32 ")",
-                      tile_key.c_str(), subquads.size(), version, depth);
-
-  for (const auto& subquad : subquads) {
-    auto subtile = tile.AddedSubHereTile(subquad->GetSubQuadKey());
+  QuadTreeIndex tree(tile, depth, quad_tree.response);
 
-    // Add to result
-    result.emplace(subtile, subquad->GetDataHandle());
-
-    // add to bulk partitions for cacheing
-    partitions.GetMutablePartitions().emplace_back(
-        PartitionsRepository::PartitionFromSubQuad(*subquad,
-                                                   subtile.ToHereTile()));
+  if (tree.IsNull()) {
+    OLP_SDK_LOG_WARNING_F(kLogTag,
+                          "QuadTreeIndex failed, hrn='%s', "
+                          "layer='%s', root='%s', version='%" PRId64
+                          "', depth='%" PRId32 "'",
+                          catalog.ToString().c_str(), layer_id.c_str(),
+                          tile_key.c_str(), version, depth);
+    return {{client::ErrorCode::Unknown, "Failed to parse QuadTreeIndex json"}};
   }
-
   // add to cache
-  repository::PartitionsCacheRepository cache(
-      catalog, settings.cache, settings.default_cache_expiration);
-  cache.Put(PartitionsRequest().WithVersion(version), partitions, layer_id,
-            boost::none, false);
-
-  return result;
+  repository.Put(layer_id, tile, depth, tree, version);
+  return get_sub_quads(tree);
 }
 
 SubQuadsResponse PrefetchTilesRepository::GetVolatileSubQuads(
```

Comment:
For volatile client `GetVolatileSubQuads` used instead of `GetSubQuads`, so need update it too

Reply:
This change will be postponed, because we don't have TileRequest for volatile layer client